### PR TITLE
backend errors

### PR DESCRIPTION
**Pull request by mjudeikis**
Return errors from the backend

### CODE REVIEW

**Comment by mjudeikis** [comment]
If this is fine, I would like to cover individual errors in the separate PR

**Comment by mjudeikis** [comment]
I'm fine with using cloudErrors here. 
Errors package is readability and cosmetics. Don't care much, it just simplification...

File: pkg/api/error.go [hunk 1]
```diff
@@ -65,6 +65,7 @@ var (
 	CloudErrorCodeInvalidLocation                    = "InvalidLocation"
 	CloudErrorCodeInvalidOperationID                 = "InvalidOperationID"
 	CloudErrorCodeDuplicateClientID                  = "DuplicateClientID"
+	CloudErrorCodeDNSCollision                       = "DNSNameCollision"
```
Comment (jim-minter) [inline pkg/api/error.go]: DuplicateDomain?

File: pkg/install/0-installstorage.go [hunk 2]
```diff
@@ -37,7 +38,11 @@ var apiVersions = map[string]string{
 }
 
 func (i *Installer) createDNS(ctx context.Context) error {
-	return i.dns.Create(ctx, i.doc.OpenShiftCluster)
+	err := i.dns.Create(ctx, i.doc.OpenShiftCluster)
+	if err != nil {
```
Comment (jim-minter) [inline pkg/install/0-installstorage.go]: make `"recordset %q already registered"` an official error in the dns package and specifically test it here

File: pkg/install/0-installstorage.go [hunk 3]
```diff
@@ -37,7 +38,11 @@ var apiVersions = map[string]string{
...
+		return api.NewCloudError(http.StatusBadRequest, api.CloudErrorCodeDNSCollision, "properties.clusterProfile.domain", "DNS creation failed with error %s.", err)
```
Comment (jim-minter) [inline pkg/install/0-installstorage.go]: leave target empty like we do in dynamic validation

Comment (jim-minter) [reply pkg/install/0-installstorage.go]: message can be something like `"The provided domain '%s' is already in use by a cluster."`

File: pkg/backend/openshiftcluster.go [hunk 4]
```diff
@@ -179,18 +172,27 @@ func (ocb *openShiftClusterBackend) heartbeat(ctx context.Context, cancel contex
 	}
 }
 
-func (ocb *openShiftClusterBackend) updateAsyncOperation(ctx context.Context, id string, oc *api.OpenShiftCluster, provisioningState, failedProvisioningState api.ProvisioningState) error {
+func (ocb *openShiftClusterBackend) updateAsyncOperation(ctx context.Context, id string, oc *api.OpenShiftCluster, provisioningState, failedProvisioningState api.ProvisioningState, backendErr error) error {
 	if id != "" {
+		log := ocb.baseLog.WithField("clusterID", oc.ID)
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster.go]: don't do this, please revert the log.Error() changes above instead

File: pkg/backend/openshiftcluster.go [hunk 5]
```diff
@@ -33,8 +33,7 @@ func (ocb *openShiftClusterBackend) try(ctx context.Context) (bool, error) {
 
 	log := ocb.baseLog.WithField("resource", doc.OpenShiftCluster.ID)
 	if doc.Dequeues > maxDequeueCount {
-		log.Errorf("dequeued %d times, failing", doc.Dequeues)
-		return true, ocb.endLease(ctx, nil, doc, api.ProvisioningStateFailed)
+		return true, ocb.endLease(ctx, nil, doc, api.ProvisioningStateFailed, fmt.Errorf("dequeued %d times, failing", doc.Dequeues))
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster.go]: if you want to do:
```
err := fmt.Errorf("dequeued %d times, failing", doc.Dequeues)
log.Error(err)
return true, ocb.endLease(ctx, nil, doc, api.ProvisioningStateFailed, err)
```
that's OK

File: pkg/backend/openshiftcluster.go [hunk 6]
```diff
@@ -179,18 +172,27 @@ func (ocb *openShiftClusterBackend) heartbeat(ctx context.Context, cancel contex
...
 			asyncdoc.AsyncOperation.ProvisioningState = provisioningState
 
 			now := time.Now()
 			asyncdoc.AsyncOperation.EndTime = &now
 
-			if provisioningState == api.ProvisioningStateFailed {
-				asyncdoc.AsyncOperation.Error = &api.CloudErrorBody{
-					Code:    api.CloudErrorCodeInternalServerError,
-					Message: "Internal server error.",
+			if provisioningState == api.ProvisioningStateFailed && backendErr != nil {
```
Comment (jim-minter) [inline pkg/backend/openshiftcluster.go]: in what circumstances can `provisioningState == api.ProvisioningStateFailed && backendErr == nil`?  Why is the `backendErr != nil` test needed?  Note: if it is nil, below `err, ok := backendErr.(*api.CloudError)` should return `nil, false`, I think?

File: pkg/backend/openshiftcluster.go (hunk 5 above)
Comment (mjudeikis) [reply pkg/backend/openshiftcluster.go]: The error will be logged by `endLease`. This way we simplify logging of the errors when the lease is ended.

**Pull request merged**